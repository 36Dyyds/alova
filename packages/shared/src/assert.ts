import { newInstance } from './function';

/**
 * 构建格式化的错误消息
 * @param prefix 错误前缀
 * @param msg 错误消息
 * @returns 格式化的错误消息
 */
export const buildErrorMsg = (prefix: string, msg: string) => `[alova${prefix ? `/${prefix}` : ''}]${msg}`;

/**
 * 自定义断言函数，表达式为false时抛出错误
 * 当传入了errorCode时，将提供链接到错误文档，引导用户改正
 * @param expression 判断表达式，true或false
 * @param msg 断言消息
 */
export const createAssert =
  (prefix = '') =>
  (expression: boolean, msg: string, errCode?: number) => {
    if (!expression) {
      throw newInstance(
        Error,
        buildErrorMsg(prefix, msg + (errCode ? `\n\nFor detailed: https://alova.js.org/err#${errCode}` : ''))
      );
    }
  };
