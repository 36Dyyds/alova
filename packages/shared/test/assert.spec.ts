import { createAssert } from '@/assert';

describe('createAssert', () => {
  test('should throw an error when expression is false with prefix', () => {
    const assert = createAssert('validation');
    expect(() => assert(false, 'Value is not valid')).toThrowErrorMatchingSnapshot();
  });

  test('should not throw an error when expression is true with prefix', () => {
    const assert = createAssert('validation');
    expect(() => assert(true, 'Value is valid')).not.toThrow();
  });

  test('should throw an error with error code and include error details URL', () => {
    const assert = createAssert();
    const errCode = 1001;
    const msg = 'Invalid configuration';
    expect(() => assert(false, msg, errCode)).toThrowErrorMatchingSnapshot();
  });

  test('should throw an error without error code', () => {
    const assert = createAssert();
    const msg = 'Something went wrong';
    expect(() => assert(false, msg)).toThrowErrorMatchingSnapshot();
  });

  test('should handle null/undefined prefix gracefully', () => {
    const assertWithNull = createAssert(null as any);
    const assertWithUndefined = createAssert(undefined);
    expect(() => assertWithNull(false, 'Test with null prefix')).toThrowErrorMatchingSnapshot();
    expect(() => assertWithUndefined(false, 'Test with undefined prefix')).toThrowErrorMatchingSnapshot();
  });
});
