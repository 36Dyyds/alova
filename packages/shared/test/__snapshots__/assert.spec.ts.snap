// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createAssert should handle null/undefined prefix gracefully 1`] = `"[alova]Test with null prefix"`;

exports[`createAssert should handle null/undefined prefix gracefully 2`] = `"[alova]Test with undefined prefix"`;

exports[`createAssert should throw an error when expression is false with prefix 1`] = `"[alova/validation]Value is not valid"`;

exports[`createAssert should throw an error with error code and include error details URL 1`] = `
"[alova]Invalid configuration

Detaild: https://alova.js.org/err#1001"
`;

exports[`createAssert should throw an error without error code 1`] = `"[alova]Something went wrong"`;
